subsys/net/lib/websocket/websocket.c

```diff
@@ -715,94 +724,157 @@ int websocket_send_msg(int ws_sock, const uint8_t *payload, size_t payload_len,
+				}
+				ctx->parser_state = WEBSOCKET_PARSER_STATE_LENGTH;
+				break;
+			case WEBSOCKET_PARSER_STATE_LENGTH:
+				ctx->masked = (data & 0x80) != 0;
+				len = data & 0x7f;
+				if (len < 126) {
+					ctx->message_len = len;
+					if (ctx->masked) {
+						ctx->masking_value = 0;
+						ctx->parser_remaining = 4;
+						ctx->parser_state = WEBSOCKET_PARSER_STATE_MASK;
+					} else {
+						ctx->parser_remaining = ctx->message_len;
+						ctx->parser_state = (ctx->parser_remaining == 0)
+							? WEBSOCKET_PARSER_STATE_OPCODE
+							: WEBSOCKET_PARSER_STATE_PAYLOAD;
+					}
+				} else {
+					ctx->message_len = 0;
+					ctx->parser_remaining = (len < 127) ? 2 : 8;
+					ctx->parser_state = WEBSOCKET_PARSER_STATE_EXT_LEN;
+				}
+				break;
+			case WEBSOCKET_PARSER_STATE_EXT_LEN:
+				ctx->parser_remaining--;
+				ctx->message_len |= (data << (ctx->parser_remaining * 8));
+				if (ctx->parser_remaining == 0) {
+					if (ctx->masked) {
+						ctx->masking_value = 0;
+						ctx->parser_remaining = 4;
+						ctx->parser_state = WEBSOCKET_PARSER_STATE_MASK;
+					} else {
+						ctx->parser_remaining = ctx->message_len;
+						ctx->parser_state = WEBSOCKET_PARSER_STATE_PAYLOAD;
+					}
+				}
+				break;
+			case WEBSOCKET_PARSER_STATE_MASK:
+				ctx->parser_remaining--;
+				ctx->masking_value |= (data << (ctx->parser_remaining * 8));
+				if (ctx->parser_remaining == 0) {
+					if (ctx->message_len == 0) {
+						ctx->parser_remaining = 0;
+						ctx->parser_state = WEBSOCKET_PARSER_STATE_OPCODE;
+					} else {
+						ctx->parser_remaining = ctx->message_len;
+						ctx->parser_state = WEBSOCKET_PARSER_STATE_PAYLOAD;
+					}
+				}
+				break;
+			default:
+				return -EFAULT;
+			}
 
-	if (buf_len >= *header_len) {
-		if (value & 0x0080) {
-			*masked = true;
-			*mask_value = sys_get_be32(&buf[2 + len_len]);
-			*header_len += 4;
+	#if (LOG_LEVEL >= LOG_LEVEL_DBG)
+			if ((ctx->parser_state == WEBSOCKET_PARSER_STATE_PAYLOAD) ||
+				((ctx->parser_state == WEBSOCKET_PARSER_STATE_OPCODE) &&
+				 (ctx->message_len == 0))) {
+				NET_DBG("[%p] %smasked, mask 0x%08x, type 0x%02x, msg %zd",
+						ctx, ctx->masked ? "" : "un",
+						ctx->masked ? ctx->masking_value : 0,
+						ctx->message_type, (size_t)ctx->message_len);
+			}
+	#endif
 		} else {
-			*masked = false;
+			size_t remaining_in_recv_buf =
+						ctx->recv_buf.count - parsed_count;
+			size_t payload_in_recv_buf =
+						MIN(remaining_in_recv_buf, ctx->parser_remaining);
+			size_t free_in_payload_buf = payload->size - payload->count;
+			size_t ready_to_copy =
+						MIN(payload_in_recv_buf, free_in_payload_buf);
+
+			memcpy(&payload->buf[payload->count],
+					&ctx->recv_buf.buf[parsed_count],
+					ready_to_copy);
+			parsed_count += ready_to_copy;
+			payload->count += ready_to_copy;
+			ctx->parser_remaining -= ready_to_copy;
+			if (ctx->parser_remaining == 0) {
+				ctx->parser_remaining = 0;
+				ctx->parser_state = WEBSOCKET_PARSER_STATE_OPCODE;
+			}
```

Comment:
The loop is missing an exit condition when the application-provided buffer is filled out, something like:
```
			if (payload->count == payload->size) {
				break;
			}
```
I've hit the issue when testing the changes with MQTT - the MQTT library reads the data in small chunks when parsing the MQTT header. There was still data in the recv buffer, but as `free_in_payload_buf` evaluated to 0, the loop spinned endlessly as it could not consume all of the data from the recv buffer.

After fixing this issue, the implementation worked fine, both with TCP and TLS underneath.

Reply:
I added test `test_recv_in_small_buffer`, which reproduce this bug, then added checking you proposed.

Reply:
Thanks